docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+
+
+## Goals
+
+* Provide a mechanism to migrate the control plane of a shoot cluster from one seed to another
+* The mechanism should support migration from a seed which is no longer reachable (Disaster Recovery)
+* The mechanism can automatically decide whether a seed is unhealthy and control planes need to be migrated to a new seed.
+* Cluster nodes which are already available and healthy should be reused after the migration completes
+* Extension controllers specify which Custom Resources need to be migrated so that the control plane can be recreated on the new seed
+* The already existing shoot reconciliation flow is reused for migration with minimum changes
+* The mechanism needs to take into account that gardenlet will be used for shoot reconciliation very soon
+
+
+## Terminology
+
+__Host Seed__ is the seed which currently hosts the control plane of a Shoot Cluster
+
+__New Seed__ is the seed to which the control plane is being migrated
+
+
+## Resources and controller state which have to be migrated between two seeds:
+
+### Secrets
+
+Gardener has preconfigured lists of needed secrets which are generated when a shoot is created and deployed in the seed. Following is a minimum set of secrets which must be migrated to the New Seed. Other secrets can be regenerated from them.
+
+- ca
+- ca-front-proxy
+- static-token
+- ca-kubelet
+- ca-metrics-server
+- etcd-encryption-secret
+- kube-aggregator
+- kube-apiserver-basic-auth
+- kube-apiserver
+- service-account-key
+- ssh-keypair
+
+### Custom Resources and state of extension controllers
+
+Gardener deploys custom resources to the Host Seed cluster which are reconciled by extension controllers. The state of these controllers and any additional resources they create is independent of gardener and must also be migrated to the New Seed. Following is a list of custom resources, and the state which is generated by them that has to be migrated.
+
+- BackupBucket: nothing relevant for migration
+- BackupEntry: nothing relevant for migration
+- ControlPlane: nothing relevant for migration
+- DNSProvider/DNSEntry: nothing relevant for migration
+- Extensions: migration of state needs to be handled individually
+- Infrastructure: terraform state
+- Network: nothing relevant for migration
+- OperatingSystemConfig: nothing relevant for migration
+- Worker: Machine-Controller-Manager related objects: machineclasses, machinedeployments, machinesets, machines
+
+This list depends on the currently installed extensions and can change in the future
+
+## Proposal
+
+### CRD on the garden cluster
+
+The garden cluster has a new CustomResource. It contains all the required data described above so that the the control plane can be recreated on the New Seed.
+The metadata is separated into two sections. The first is generated by the Gardener Controller Manager and then either used to generate new resources (e.g secrets) or data in it is directly deployed to the Shoot's control plane on the New Seed.
```

Comment:
Maybe unrelated to the migration itself, but when I read "generate secrets" here and see the below data, then I see sensitive information outside of secrets, i.e. these are not encrypted. That's a problem/not safe.

Reply:
Something we considered during the discussions.
We could either keep them as secrets and only have references in the SeedState or keep the data fields encrypted in the ShootState. 
But we can look more into the exact handling of the sensitive data after first implementing migration between working seeds (currently our first milestone).

Reply:
Agreed, but eventually the data should be moved to secrets. Its unrelated to the migration topic, but still important to keep in mind.

Reply:
The problem with secrets/configmaps is that project members have full access to them, while `ShootState` is new resource that is(will be) designed not to be provided to the end-users.

I am also for sensitive data to be put into secrets and configurations into configmaps, but they need to be isolated from the end users, which can be achieved via custom authorization or stored into dedicated namespace different from the project namespace.

Reply:
Yes, the reason why we didn't want to store it in secrets was that we wanted the `ShootState` objects in the project namespaces, and here end-users are allowed to read secrets. What we actually suggested was to put the sensitive data directly in the `ShootState` and simply encrypt this object in etcd, too.
Similarly, what we also should do as well is to encrypt the `ControllerRegistration` objects as they may also contain sensitive data for the extension controllers. OK?

Reply:
Ah, I totally forgot. The API server configuration allows to specify the resource kinds that shall be encrypted, right? I.e. we can decide freely what we like to encrypt. Great, thanks!

Reply:
Yes, however, we will need to give the gardener-apiserver a etcd encryption configuration as well, not only the kube-apiserver (because it doesn't know/recognize our objects).